### DISCUSSION TITLE
Traceability of data on invoke

### DISCUSSION THREAD

**Discussion by MaheshkumarSundaram** [bpmnServer/bpmn-server#202] [Q&A] [answered]
Hi @ralphhanna,

I would like to clarify something specific for User Tasks. I use `bpmn-client` from my react front-end to invoke user tasks.

Now, I can invoke like `await client.engine.invoke(matchCriteria, output)`;

Say, my output for all user tasks is structured like `{ key1: 'value1', key2: 'value3' }`.

This output is merged into instance's data. **But in the next User Task, these values would be replaced by that particular new output**.

But I would like to have a traceability of the data produced by User Tasks without one overwriting the other. How to achieve this?

Thank you.

**Comment by ralphhanna** [comment]
Hello @MaheshkumarSundaram 
The intent of `instance.data` is to hold data elements common to the workflow
We also have `item.vars` to hold data elements specific to the item (instance of a specific node)

To store values in `item.vars` : 
```ts
  item.vars = item.input;
  item.input = {}; // so instance.data will not save the input as usual
```

Please let me know if this meets your request?

Thanks

**Reply by MaheshkumarSundaram** [reply]
@ralphhanna But how to assign the data to `item.vars` while invoking? The data is provided by the users when completing a user task dynamically as required.

**Comment by ralphhanna** [comment]
Sorry, using ‘validate’ trigger
Sent from my iPhone

On Jun 11, 2024, at 3:43 PM, Maheshkumar Sundaram ***@***.***> wrote:

﻿

@ralphhanna<https://github.com/ralphhanna> But how to assign this while invoking? The data is provided by the users when completing a user task dynamically as required.

**Reply by MaheshkumarSundaram** [reply]
@ralphhanna So sorry, I am failing to understand. Could you please give me an example of how would I call this using `bpmn-client`

**Comment by ralphhanna** [comment]
You need to modify the model:

![image](https://github.com/bpmnServer/bpmn-server/assets/11893416/39682433-91a1-49e7-bf5e-bf94b0276651)


![image](https://github.com/bpmnServer/bpmn-server/assets/11893416/3a69a412-f8d7-4fe8-8d41-a0ab7a3b023a)

**Reply by MaheshkumarSundaram** [reply]
Sorry @ralphhanna. But how does using `validate` help when I want to capture the data on invocation of a user task from `bpmn-client`? The data is not provided through forms directly. It's created by the application depending on the states in my react application. I'll only know the data before invocation.

**Comment by ralphhanna** [comment]
If I understand correctly, `invoke` executes `validate` before applying the data.
Please get latest release of bpmn-server.

https://bpmnserver.github.io/docs/execution/#engineinvoke

**Comment by ralphhanna** [comment, marked as answer]
Correction to earlier note, the correct syntax is as follows:
```js
item.vars=item.input;
delete item.data.needsRepairs; // if you want to remove values from instance.data
```
![image](https://github.com/bpmnServer/bpmn-server/assets/11893416/2d0260e9-a951-4e4d-a62b-1d344dcd34a9)

**Reply by MaheshkumarSundaram** [reply]
@ralphhanna Thank you so much. This works fine 👍

Is there any plan to streamline handling UserTask's data in near future? Although this solution works, using validate for the above use case seems quite ambiguous as it's not validating anything essentially. Sorry if am nitpicking but just a thought.

It seems this can be done also using `end` Execution Listener. I just tested this. I prefer using the execution listener to do this as it seems right rather than `validate`. What do you think? Thanks.

**Comment by ralphhanna** [comment]
Hi @MaheshkumarSundaram 
`validate` is done before completing the task and can return errors to reject the operation, while `end` can not reject the operation.
Not sure why it is not validating the user input. 
Please give more details
Thanks

**Reply by MaheshkumarSundaram** [reply]
@ralphhanna, Sorry if I confused you. `validate` is fine. But for my use case, I don't need `validate` as I am not validating anything. So, I said I'll just use `end` listener.
